io-classes/src/Control/Monad/Class/MonadSTM/Internal.hs

```diff
@@ -1257,3 +1261,40 @@ instance MonadSTM m => MonadSTM (ReaderT r m) where
 writeTMVar' :: STM.TMVar a -> a -> STM.STM ()
 writeTMVar' t new = STM.tryTakeTMVar t >> STM.putTMVar t new
 #endif
+
+
+-- | Apply @f@ with the content of @tv@ as state, restoring the original value when an
+-- exception occurs
+withTMVar ::
+     (MonadSTM m, MonadThrow.MonadCatch m)
+  => TMVar m a
+  -> (a -> m (c, a))
+  -> m c
+withTMVar tv f = withTMVarAnd tv (const $ pure ()) (\a -> const $ f a)
```

Comment:
Nitpick: Shouldn't this be called `modifyTMVar`? From `base`:
```haskell
withMVar :: MVar a -> (a -> IO b) -> IO b
modifyMVar :: MVar a -> (a -> IO (a, b)) -> IO b
```
(also, the order of arguments in the tuple should be flipped :sweat_smile:)

Reply:
The [`stm`](https://hackage.haskell.org/package/stm-2.5.3.1/docs/Control-Concurrent-STM-TMVar.html) package doesn't expose `withTMVar` that's why we don't expose it too.

Reply:
This function is now gone